GVFS/GVFS.Common/Actions/AbstractObjectCacheAction.cs

```diff
@@ -0,0 +1,184 @@
+﻿using GVFS.Common.FileSystem;
+using GVFS.Common.Git;
+using GVFS.Common.Tracing;
+using System;
+using System.IO;
+
+namespace GVFS.Common.Actions
+{
+    public abstract class AbstractObjectCacheAction : IAction

```

Comment:
Having read through the code, I'm not sure that we should have one class that manages both "background tasks that generate new content (prefetch, midx, etc)" and "background tasks that perform cache cleanup". I can see why you're trying to share their implementations and that makes sense - but my worry is that their goals and tradeoffs will be different and that mismatch will create bugs.

The fact that it's so hard to even name these classes is a good clue that they are doing too much. The best I can come up with is "BackgroundTask", "CleanupJob", etc, all of which are overly broad.

An alternative approach could be: create a "BestEffortBackgroundQueue" (as opposed to the required one that we have today) that allows any code to queue up optional work, and maintains a single thread (or pool of threads) to process that queue. Then you can have classes like "BackgroundPrefetch", "MidxGenerator", "LocalCacheCleanup", etc that can all have their own business logic and custom implementation tradeoffs, but share just the mechanism of how to run a thing in the background, and how to cancel it when unmounting.

Reply:
(I'm using "actions" for this discussion as I don't have a better word right now.)

You're right that this name is too generic, especially because it is doing a lot of logic around starting a `GitProcess` that it can terminate during `Stop()`. That should be the core of this class. As such, I should [move the pack-file modified time logic](https://github.com/Microsoft/VFSForGit/pull/493/files/da9e1b29d0fc0077420284304bf0c69f91b5f649#r233604138) into another class. The other logic around the `ObjectCacheLock` is important for protecting actions in different mount processes from fighting.

In hindsight, the `PrefetchAction` also runs `git` processes, and could benefit from this machinery (it doesn't extend this class). It would start multiple processes during the course of the action, but it could use `RunGitCommand` to do that in a safe way. This means we can remove `IAction` and have every action extend this class and rename it something like `GitObjectCacheMaintenanceStep`. We would want to split out an interface later if we decide to use this mechanism for things like the shared object size database, but YAGNI says we shouldn't do it now.

> ...that allows any code to queue up optional work...

The current post-fetch job works this way, so would be easy to replace the old `LaunchPostFetchJobIfIdle()` with a `QueuePostFetchJobs()` method. We could then separate the regularly-scheduled maintenance into `Timer`s that have different periods and all they do is queue a job. This sounds a bit like a background job agent.

We could then split this into:

* `BackgroundQueue` takes `QueueJob` requests and adds them to a `ConcurrentList`. It also holds a `Thread` object, that pulls jobs off of this list and runs them until the list is empty. After adding the job to the list, check if the thread is active. If the thread is active, then it will pick up the job when it can. If the thread is inactive, launch a new thread to empty the queue.
* `MaintenanceScheduler` creates a list of `Timer` objects that fire at different periods to queue jobs. The background prefetch job would run every 70 minutes. The loose object cleanup job could run less frequently.
* The `InProcessMount` class can listen for post-fetch job requests and queue those jobs according to the message contents.

I'm still not thrilled that the background prefetch sends a message over the named pipe to queue more background activity, but at least we consolidate the thread where we do the work.

The final thought I'm worried about is the "best effort" nature of this work. This is really for the actions to decide (hence why I dropped it from the class name), but we should also be careful if we actually want these actions to all take the same lock. For instance, can we write a commit-graph while we clean up loose objects? Probably. We don't want to do write two commit-graphs at the same time, and we don't want to write a multi-pack-index at the same time we may delete a pack-file. This means we should be more granular about the `ObjectCacheLock` that I'm currently managing in the base class.

Reply:
Wondering if it would simplify things if we had the service coordinate some of this since there is only one service that is running?  Mount processes could send requests to the service to run something but it would be the service's job to keep track of what is running and when?

Reply:
@kewillford The service is a nice singleton to help coordinate, but is it better than using a file-based lock? The stumbling block I am trying to overcome is what happens when a mount process suddenly dies while holding a "service lock"? The file-based lock opens up because the handle is dropped, right? Or maybe I'm misunderstanding what you are proposing.

Reply:
I've thought about that as well but this would introduce a new failure mode for a mount process. Currently, the GVFS.Service is entirely optional for a mount process to function correctly. Moving some of the mount process's logic over to the service means that now we absolutely require the service to be running, or else you lose things like midx generation.

Also that move wouldn't change any of the design discussions we're having in this PR. It just changes which process that code runs in.

Reply:
@derrickstolee 
> The final thought I'm worried about is the "best effort" nature of this work. This is really for the actions to decide (hence why I dropped it from the class name)

I disagree on that point. We have an existing background queue, and its special quality is that it is for absolutely required work, and that's why the queue is backed by disk. It is resilient to crash, power loss, etc, because that work cannot be lost.

The nature of this new queue is that it is best effort, meaning that it does not need to be backed by disk. A caller of this queue cannot decide that its action is required - if that is true, it is using the wrong queue.

Reply:
Thanks, both.

> The nature of this new queue is that it is best effort, meaning that it does not need to be backed by disk. A caller of this queue cannot decide that its action is required - if that is true, it is using the wrong queue.

Hopefully the new, very-specific name for the queue helps clarify the purpose of the queue, and the fact that it is "best-effort" is understood for the steps that are run by the queue.